state/cleanup.go

```diff
@@ -531,6 +574,123 @@ func (st *State) cleanupDyingUnit(name string, cleanupArgs []bson.Raw) error {
 	}
 }
 
+func (st *State) scheduleForceCleanup(kind cleanupKind, name string) {
+	deadline := st.stateClock.Now().Add(forceTimeout)
+	op := newCleanupAtOp(deadline, kind, name)
+	err := st.db().Run(func(int) ([]txn.Op, error) {
+		return []txn.Op{op}, nil
+	})
+	if err != nil {
+		logger.Warningf("couldn't schedule %s cleanup: %v", kind, err)
+	}
+}
+
+func (st *State) cleanupForceDestroyedUnit(unitId string) error {
+	unit, err := st.Unit(unitId)
+	if errors.IsNotFound(err) {
+		logger.Debugf("no need to force unit to dead %q", unitId)
+		return nil
+	} else if err != nil {
+		return errors.Trace(err)
+	}
+
+	// If we're here then the usual unit cleanup hasn't happened but
+	// since force was specified we still want the machine to go to
+	// dead.
+
+	// Destroy all subordinates.
+	for _, subName := range unit.SubordinateNames() {
+		subUnit, err := st.Unit(subName)
+		if errors.IsNotFound(err) {
+			continue
+		} else if err != nil {
+			logger.Warningf("couldn't get subordinate %q to force destroy: %v", subName, err)
+		}
+		opErrs, err := subUnit.DestroyWithForce(true)
+		if len(opErrs) != 0 || err != nil {
+			logger.Warningf("errors while destroying subordinate %q: %v, %v", subName, err, opErrs)
+		}
+	}
+
+	// LeaveScope on all of the unit's relations.
+	relations, err := unit.RelationsInScope()
+	if err == nil {
+		for _, relation := range relations {
+			ru, err := relation.Unit(unit)
+			if err != nil {
+				logger.Warningf("couldn't get relation unit for %q in %q: %v", unit, relation, err)
+				continue
+			}
+			err = ru.LeaveScope()
+			if err != nil {
+				logger.Warningf("unit %q couldn't leave scope of relation %q: %v", unitId, relation, err)
+			}
+		}
+	} else {
+		logger.Warningf("couldn't get in-scope relations for unit %q: %v", unitId, err)
+	}
+
+	// Detach all storage.
+	err = st.forceRemoveUnitStorageAttachments(unit)
+	if err != nil {
+		logger.Warningf("couldn't remove storage attachments for %q: %v", unitId, err)
+	}
+
+	// Mark the unit dead.
+	err = unit.EnsureDead()
+	if err == ErrUnitHasSubordinates || err == ErrUnitHasStorageAttachments {
+		// In this case we do want to die and try again - we can't set
+		// the unit to dead until the subordinates and storage are
+		// gone, so we should give them time to be removed.
+		return err
+	} else if err != nil {
+		logger.Warningf("couldn't set unit %q dead: %v", unitId, err)
+	}
+
+	// Set up another cleanup to remove the unit in a minute if the
+	// deployer doesn't do it.
+	st.scheduleForceCleanup(cleanupForceRemoveUnit, unitId)
```

Comment:
Maybe in this case we really to want to wait a bit to save thrashing on error in --force

Reply:
I think scheduling this immediately is ok - if the cleanup is failing because of storage or subordinates it will have exited before now. This will run at most once - if it happens, the cleanup will succeed and be removed, so it can't cause thrashing.